### PR TITLE
Vrapper Eclipse keys

### LINKED ISSUE #108: Automatically restore keybindings when vrapper is disabled
It would be nice if vrapper could save the key bindings that users have to disable to get their preferred vim key mappings working. When the toggle vrapper button is released, it should restore the key bindings. 

E.g. if .vrapper configuration file has the following to map CTRL-A to beginning-of-line 
imap <C-a> 0

Then while vrapper was enabled, user disables the key binding for CTRL-A in eclipse to mean "select all". After vrapper is disabled, the key bindings for CTRL-A to "select all" is restored in eclipse. 

The implementation might be a bit crude, but any solution that provides the goal of restoring key bindings that were disabled for vrapper to work is the goal.

### PR DESCRIPTION
**Pull request by albertdev**
So this is a rather big one and I currently don't feel like updating the documentation site, so I thought I could make a pull request with info to garner feedback. Help with documenting this is welcome!

---

First things first: a command which recovers the focus for the current editor. As documented in #435 this happens from time to time. I've been testing with the following settings:
issue_435_key_config

Basically press <kbd>Esc</kbd> anywhere in the Eclipse window and the editor will be activated. Obviously, that key would normally be "eaten" so I actually forward whatever key is bound to the command to Vrapper. This command cannot be disabled by toggling Vrapper so I recommend only using it for the current usecase.

---

Next up is one of our biggest pain points: Vrapper is mainly a "hack" by listening for keys sent to the current text editor. Since Eclipse would eat certain keys to trigger Eclipse commands we would tell people to unbind shortcuts so Vrapper could finally see them.

Worry no more. The commandline / search mode now captures Home / End / Next word / Previous Word / Cut / Copy / Paste by default, and everything else should be a matter of configuration through the General > Keys window:

issue_108_key_config

The keys shown here will take precedence on whatever Eclipse has normally bound to them. If you want to add another, just select one of the existing "Send to Vrapper" shortcuts and press **Copy Command**. You can then bind this empty one.

And just to show that I didn't have to unbind anything, here's a screenshot for what my <kbd>Home</kbd> key is currently bound to:

issue_108_key_config_still_bound

Finally, do notice the **"When"** column and dropdown box because this is where the magic happens. There are 6 values to be used with Vrapper:

* **Vrapper Active** will accept that key when Vrapper is enabled and an editor with Vrapper inside of it is focused. Switching to a graphical editor or a View will disable this key.
* **Vrapper Commandline Opened** (works for Search, Command, Sneak, ... modes), **Vrapper Insert mode Active**, **Vrapper Normal mode Active**, **Vrapper Visual mode Active** all accept keys when Vrapper is enabled and an editor with Vrapper inside is focused. Switching away will disable this key. Also, any key bound for this mode will override whatever is bound to **Vrapper Active**.
* **Vrapper Enabled on a View** will accept that key when you shifted the focus to an Eclipse view. No keys will be sent to Vrapper, but it is reserved for a different kind of shortcut. More on that below.

Additionally Vrapper adds some clutter at the end of the list with the name **ZZ Vrapper's Dummy Contexts**, but those shouldn't be used.

Anyway, to summarize: enable Vrapper and all those keys work, disable it and everything will works as a plain Eclipse install (given you didn't alter the default binds). We should thus be able to close #108.

---

And now for the final feature: we have had requests like #173 of people who would like to use the same modal input in Eclipse's views like the Project Explorer or Search. Sadly, there's no simple base class which Eclipse views extend from like it is the case with Eclipse editors **and** which gives us all the necessary listeners to react to keyboard input.

As a workaround I have come up with something else:

issue_173_key_config

Basically you have 4 commands (one for each direction) which you can bind to HJKL or any other shortcut. Seeing how these are enabled when **Vrapper Enabled on a View** accepts it, you can then navigate through Package Explorer pressing J / K and Vrapper will translate these into Up / Down key presses.

As a fallback, the code also tries to detect if the focus is on a text field:

issue_173_focus_textfield

In this case the code will see if your shortcut is a plain character (i.e. not triggered by <kbd>F1</kbd> or <kbd>Ctrl+F</kbd> or something) and inserts it into the text field.

If there are any other cases you can always disable Vrapper (the shortcut should stop working so the default behavior is restored), or open a bug ticket to request us to handle it as another special case.

### CODE REVIEW

**Comment by keforbes** [comment]
Wow! Way to go @albertdev!  Just a couple questions:

> Additionally Vrapper adds some clutter at the end of the list with the name ZZ Vrapper's Dummy Contexts, but those shouldn't be used.

Is there any reason to keep that Dummy Context or is it still used internally?

> The commandline / search mode now captures Home / End / Next word / Previous Word / Cut / Copy / Paste by default

Are those the only default changes or did you add any other defaults?  I'm thinking <kbd>Home</kbd> and <kbd>End</kbd> could probably be added to the `Vrapper Active` since our behavior is the same as the default, it just gives Vrapper knowledge of those keys being pressed.

**Comment by albertdev** [comment]
>> Additionally Vrapper adds some clutter at the end of the list with the name ZZ Vrapper's Dummy Contexts, but those shouldn't be used.
>
> Is there any reason to keep that Dummy Context or is it still used internally?

It's part of the magic. In #778 I gave an overview of how the contexts are stacked to give a higher "stack depth" and hence a higher score to Vrapper commands. Those intermediate contexts are just that and cannot be used.

BTW: we could change the display name if necessary.

>> The commandline / search mode now captures Home / End / Next word / Previous Word / Cut / Copy / Paste by default
>
> Are those the only default changes or did you add any other defaults? I'm thinking Home and End could probably be added to the Vrapper Active since our behavior is the same as the default, it just gives Vrapper knowledge of those keys being pressed.

Ah, right, I forgot to write that currently it doesn't include any default key binds although it **could** (note that in the screenshots above I have added those binds myself, the User columns contains a capital 'U' as an indication).

BTW: Home and End are not identical. I know the JDT plugin has a useful trick where a "rename refactor operation" will limit the Home and End keys to just the word / function name being renamed. I actually like that.

Now, I was originally planning to map a bunch of Ctrl / Alt / Shift + A, B, C, ... keys to that single Vrapper command so that every one of them is sent to Vrapper if you install / update the plugin. What made me concerned though is that everyone out there has got a pretty stable configuration at this moment, where they can trigger their favorite Eclipse combinations and invoke the Vrapper commands they need when they unbound the Eclipse stuff. If we add out-of-the-box keybinds the entire default would flip upside down and maybe disrupt people's workflow.

I was therefore thinking of salvaging this little hidden feature:

keyboard_scheme

Vrapper has been adding this scheme "Vim's key bindings" since 2010 and before, but it's always been useless. What we could do is then add all combinations of shortcuts to that and keep our current setup. The install instructions could then ask people to switch to that scheme if they want to (after having taken a backup <b>should they want to go back&nbsp;!!!</b>).

&nbsp;

At this point you might be wondering when I'm coming back to the commandline stuff in your original question: it works without default key binds. There I'm using another trick: when the shortcut for the command `org.eclipse.ui.edit.paste` is triggered, it will ask an internal Eclipse service for a handler to do the actual work. I've found a way to register an alternative handler which gets executed instead of the usual "paste in the current editor" code.

I chose to limit this technique to just the mentioned commands 1) because when the user *wants* the original behavior he cannot choose, the priority is hardcoded in our plugin.xml, and 2) because it is hardcoded in the plugin.xml I would need to repeat the same boilerplate for each possible Eclipse command (an impossible job with all the Eclipse plugins out there), 3) because those commands have a very clear mapping to what our command line does, a command like "Extend selection" will be hard to implement.

BTW: I just realized that "Select all" (commonly bound to <kbd>Ctrl + A</kbd>) is not yet handled in the command line. This is another one which could be easy to implement but currently isn't.

---

OK, that was another wall of text, but I believe this stuff is worth discussing in detail before we surprise our users. 

TL;DR:

1. Currently there are no default key binds. We can add default key binds but it would impact the user, so maybe we could go for the optional scheme instead.
1. If there are any commands which we do want to foresee a default implementation for using the second technique then we could.

**Comment by albertdev** [comment]
I also forgot that <kbd>Home</kbd> and <kbd>End</kbd> might not need rebinding if I finalize one of my other prototypes, more specifically the one in this branch: https://github.com/albertdev/vrapper/tree/EXPERIMENTAL-eclipse-motion-textobjects

Its main feature is that it can react to any Eclipse command implementing a motion or text object (anything manipulating the selection) as long as the command id is registered as one of 3 categories: a command (which resets the selection and Vrapper modes), a motion (which can be used in normal mode *and* one of the visual modes) or a textobject (basically Vrapper goes to visual mode if not already so, then syncs its selection state with whatever Eclipse chose to do).

As a result the default Home and End commands would be OK to execute.

**Comment by albertdev** [comment]
As promised I've added a bunch of shortcuts to the plugin.

For starters I've added default shortcuts to all Vrapper commands visible in the Default keyboard scheme:

issue_108_config_defaultscheme

This might look odd thanks to the <kbd>F15</kbd> in each sequence, but it is my best shot at defining a shortcut which you can't (easily) trigger while still having the "When" column properly filled in. If you want to customize the keys, just replace the key sequence and you're good to go. If you want more keys, then simply click the `Copy Command` button.

Finally I've added a boatload of shortcuts to the "Vim's key bindings" scheme:

issue_108_config_vimscheme

Picking this scheme will now make sure that most if not all Control-something shortcuts are sent directly to Vrapper. I've also mapped <kbd>Home</kbd> (with and without Control), <kbd>Esc</kbd> and the H/J/K/L keys for the package explorer since I figure that people can try it out if they switch schemes.

(Note that the "F15, ..." keys are still in there, but I found it not worth investigating to leave them out of the "Vim's key bindings" scheme since they're unlikely to be a problem.)

---

At this point I think it's stable enough to actually merge it into our master branch and create a build, but it could use some feedback on whether it works in Linux, Eclipse 3.7, migrated workspaces with custom shortcuts, etc.

**Comment by albertdev** [comment]
Oh yeah, since I've added <kbd>Control</kbd>+<kbd>Home</kbd> this should cover #801 as well.

**Comment by albertdev** [comment]
Aaaaaand I just found a bug in the H/J/K/L View keys. :-(

Eclipse has a nifty view called "Debug Shell" (used to be Display) where you can enter and evaluate expressions during debugging. Turns out that the workaround for textfields which I described above doesn't properly work: it inserts the character *after* the caret, thus turning your typed text into gobbledygook. That will need to be fixed before merging...

**Comment by seflue** [comment]
What is the status on this one?

**Comment by albertdev** [comment]
@seflue I have been using it for a while in January and managed to iron out the issues listed above.

However, since I don't use Eclipse no more (it's nearly always Visual Studio now, doing .NET stuff) it means that I have not fully tested it on the newest Eclipse.

Maybe we should simply merge it, push it to the Unstable repository and let other people chime in with their experiences.

@keforbes What do you think?

**Comment by albertdev** [comment]
Ah yes, one thing of note: my pull request does not modify the website, yet this new feature can definitely use some documentation.

Basically, if you don't change the "Key scheme" then Vrapper will not react to <kbd>Control</kbd>+ *X* just like it was in the past, so some people might want to learn how to change that Key scheme.

Other people might want to go wild with the new binding options that this PR offers, but some examples or a step-by-step guide will then obviously be appreciated.

**Comment by keforbes** [comment]
> Basically, if you don't change the "Key scheme" then Vrapper will not react to Control+ X just like it was in the past, so some people might want to learn how to change that Key scheme.

Does this mean all default <kbd>Ctrl</kbd> keybindings will be disabled unless the user changes the Key Scheme?  I don't want to "break" an existing install because of the upgrade.  No one will know they need to manually change the Key Scheme to get their <kbd>Ctrl</kbd> keys back.

Can we change the Key Scheme to Vrapper at install time so those keys continue working?  Users can always change away from the Vrapper key scheme if they really want a different scheme.  I guess in my opinion, anyone who installs Vrapper prefers vim key bindings and they probably want this set by default so everything continues working as expected.

I'm tempted to pull this in so people can try it, but it'll be a while before I can update the documentation on how to use this feature.  I want the least amount of disruption for everyone not following the changelog.

**Comment by seflue** [comment]
Changing the Key Scheme to Vrapper at install time would be convenient but to try things out I think it would be ok to just add it to the documentation. I'm eager to try this PR out - it is annoying to unbind all Eclipse Standard Keybindings to avoid collision with standard Vim/Vrapper bindings.

A nice feature would be toggling between "Default" and Vrapper key scheme when switching between normal and Vrapper mode.

**Comment by albertdev** [comment]
>> Basically, if you don't change the "Key scheme" then Vrapper will not react to Control+ X just like it was in the past, so some people might want to learn how to change that Key scheme.
>
> Does this mean all default Ctrl keybindings will be disabled unless the user changes the Key Scheme? I don't want to "break" an existing install because of the upgrade. No one will know they need to manually change the Key Scheme to get their Ctrl keys back.

Ah, it seems we have a misunderstanding. Installing a version of Vrapper with the changes in this PR should not affect the keybinding of existing installs because I did *not* bind any <kbd>Ctrl</kbd> + X keys in the Default keyboard scheme. If somebody unbound a bunch of Eclipse's <kbd>Ctrl</kbd> keys to be able to send them to Vrapper then those will remain unbound. On the other hand, if somebody bound a bunch of keys which have nothing to do with Vrapper then they will not be touched.

My remark was simply that new installs of Vrapper will still behave like Vrapper always has: Eclipse handles some keys which are then never seen by Vrapper. This isn't bad in itself (because it means that the behaviour is perfectly consistent with older versions), it's just that we might still get support questions saying that "Vrapper does not handle some keys". We then have to tell them that they have to change 1 setting more (the Keyboard scheme) will be far more acceptable than "unbind all your Eclipse keys".

The documentation could have a screenshot of how to change the Keyboard Scheme.

> Can we change the Key Scheme to Vrapper at install time so those keys continue working? Users can always change away from the Vrapper key scheme if they really want a different scheme. I guess in my opinion, anyone who installs Vrapper prefers vim key bindings and they probably want this set by default so everything continues working as expected.

No, because there's no event which gets triggered when Vrapper is installed, and I wouldn't like it when the plugin tries to change the keyboard scheme if it detects that it runs for the first time (which is hard when we never actually tracked it, so existing installs might be affected as well).

> I'm tempted to pull this in so people can try it, but it'll be a while before I can update the documentation on how to use this feature. I want the least amount of disruption for everyone not following the changelog.

No defaults are changed so people don't have to keep track of the changelog. I was talking only about new users.

**Comment by albertdev** [comment]
@seflue 

> Changing the Key Scheme to Vrapper at install time would be convenient but to try things out I think it would be ok to just add it to the documentation. I'm eager to try this PR out - it is annoying to unbind any particular Eclipse Standard Keybinding manually which collides with a standard Vim/Vrapper binding.

Not something I would recommend, see my earlier comment.

> A nice feature would be toggling between "Default" and Vrapper key scheme when switching between normal and Vrapper mode.

Could it be that there is another misunderstanding?

It's not about the Key Scheme. The Key Scheme only defines shortcuts. The real "magic" happens in the "Send to Vrapper" command combined with the condition in the "When" field and that works no matter what Key Scheme you pick.

To give another example: suppose you bind a shortcut <kbd>Ctrl</kbd>+<kbd>O</kbd> to the "Send to Vrapper" command with a "When" condition "Vrapper Active". However, Eclipse also has a <kbd>Ctrl</kbd>+<kbd>O</kbd> shortcut for Quick Outline (if you didn't remove it earlier). With this PR, both can exist at the same time. When you disable Vrapper by clicking the button in the toolbar, pressing <kbd>Ctrl</kbd>+<kbd>O</kbd> will show the Quick Outline view. When you enable Vrapper, pressing <kbd>Ctrl</kbd>+<kbd>O</kbd> will work just like Vim.

The Default key scheme contains some Vrapper shortcuts as an example for the people who want to setup everything manually (and as I said above, I didn't change it any further so as to not break existing installs). The "Vim's key bindings" key scheme [contains keys bound to every combination](https://user-images.githubusercontent.com/1451362/50428802-8e8c2800-08ba-11e9-9c42-6967ad57fdda.PNG) which seemed interesting to me so that people don't have to set it up by hand if they want the "work like Vim" mode.

**Comment by albertdev** [comment]
I have rebased my changes so that a clean "Merge pull request" could be done.

**Comment by keforbes** [comment]
Thanks for clarifying those misunderstandings.  I'll pull this in.

**Comment by keforbes** [comment]
@seflue, I've pulled in this change and pushed a new build to our unstable update site (0.75.20190521).  Go ahead and give it a try and let us know if you run into any problems.  Thanks for poking us to get this committed! :smile:

**Comment by seflue** [comment]
Thanks guys for pulling this in, I will try it out as soon as possible and 
I'll be glad to give you some feedback.

Am 22. Mai 2019 02:23:51 schrieb keforbes <notifications@github.com>:
